design/one-pager-resource-connectivity-mvp.md

```diff
@@ -0,0 +1,653 @@
+  [GKE cluster external resource]'s `.ipAllocationPolicy.useIpAliases` and
+  `.ipAllocationPolicy.createSubnetwork` fields to ensure an appropriate VPC
+  native subnet is automatically created and used.
+
+To add a generous amount of flexibility Crossplane could:
+
+* Add support for creating a new VPC network via a Crossplane VPC `Network`
+  managed resource.
+* Add support for creating a new VPC subnetwork via a VPC `Subnetwork` managed
+  resource.
+* Add support for `CloudSQLInstance` managed resources and their resource
+  classes to specify the external resource's `.ipConfiguration.privateNetwork`,
+  and specify the external name of a `Network` managed resource.
+* Add support for `GKECluster` managed resources to specify the [GKE cluster
+  external resource]'s `.network` and `.subnetwork` fields.
+* Add support for `GKECluster` managed resources and their resource classes to
+  specify the [GKE cluster external resource]'s `.network` field and
+  `.subnetwork` fields, and specify the external names of a `Network` and a
+  `Subnetwork` managed resource.
+* Use the `GKECluster` managed resource's existing support for setting the
+  [GKE cluster external resource]'s `.ipAllocationPolicy.useIpAliases` field.
+
+High-fidelity Crossplane managed resource representations of the aforementioned
+network and subnetwork external resources would look as follows:
+
+```yaml
+---
+apiVersion: vpc.gcp.crossplane.io/v1alpha1
+kind: Network
+metadata:
+  namespace: crossplane-system
+  name: example
+spec:
+  nameFormat: mycoolname
+  description: A really cool VPC network
+  # IPv4Range puts the network in subnetwork-less 'legacy mode'
+  IPv4Range: 10.0.0.0/8
+  # The documentation says this puts the VPC network in auto-create subnet mode.
+  # In my experiments it seemed to create the VPC in legacy mode instead.
+  autoCreateSubnetworks: true
+  peerings:
+  - name: peerwithsomeothercoolnetwork
+    network: someothercoolnetwork
+    autoCreateRoutes: true
+    exchangeSubnetRoutes: true
+  routingConfig:
+    routingMode: REGIONAL
+```
+
+```yaml
+---
+apiVersion: vpc.gcp.crossplane.io/v1alpha1
+kind: Subnetwork
+metadata:
+  namespace: crossplane-system
+  name: example
+spec:
+  nameFormat: mycoolname
+  description: My cool VPC subnetwork
+  network: projects/coolproject/global/networks/coolestvpc
+  ipCidrRange: 192.168.0.0/24
+  region: us-central-1
+  privateIpGoogleAccess: true
+  secondaryIpRanges:
+  - rangeName: pods
+    ipCidrRange: 10.0.0.0/8
+  - rangeName: services
+    ipCidrRange: 172.16.0.0/16
+  enableFlowLogs: true
+  logConfig:
+    enable: true
+    flowSampling: 0.5
+    metadata: INCLUDE_ALL_METADATA
+    aggregationInterval: 5-min
+```
+
+Putting this all together, the infrastructure administrator would configure the
+following to ensure that when an app operator created a `MySQLInstance` claim
+and `KubernetesCluster` claim the two would have connectivity:
+
+```yaml
+---
+# A Network managed resource.
```

Comment:
I may have missed this somewhere in the document, but does this proposal mention resource classes *for* network resources at all? For example, let's say an app developer needs to create a new `Subnetwork` as part of their application. This gets difficult because will their be a general claim type for subnetworks that is portable? Can an admin create a GCP `SubnetworkClass` (assuming strongly typed resource classes) that specifies things like provider and name format and can be referenced by a claim to create a new `Subnetwork`? Is this something that we are deferring to a later discussion?

Reply:
Good catch. I attempted to gloss over that detail in this design, because it's not obvious how such things would be useful until such time as we're able to model dependencies between resource claims. Put otherwise:

* I can't imagine an application developer using an abstract VPC network resource claim to ask Crossplane to create an abstract VPC network for the sake of having a VPC network alone.
* I _can_ imagine an application developer using an abstract VPC network resource claim as a home for other abstract resource claims (clusters, databases, etc), but this takes us down the path of cross-cloud networking that we're not ready to attack just yet.